File: src/agent/loop.py
"""Agent execution loop — call LLM with tools, execute tool calls, repeat."""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass, field
from typing import Any

from src.agent.profile import AgentProfile
from src.agent.tools import ToolRegistry
from src.llm.gateway import LLMGateway

logger = logging.getLogger(__name__)


@dataclass
class ToolCallRecord:
    """Record of a single tool call within an agent run."""

    tool_name: str
    arguments: dict[str, Any]
    result: Any
    iteration: int


@dataclass
class AgentResult:
    """Result of an agent run."""

    status: str  # "completed", "max_iterations", "error"
    final_message: str = ""
    tool_calls: list[ToolCallRecord] = field(default_factory=list)
    iterations: int = 0
    error: str | None = None


class AgentLoop:
    """Tool-use loop: call Claude -> execute tool calls -> feed results back -> repeat."""

    def __init__(
        self,
        llm: LLMGateway,
        tool_registry: ToolRegistry,
    ):
        self.llm = llm
        self.tool_registry = tool_registry

    def run(
        self,
        profile: AgentProfile,
        user_message: str,
        *,
        user_id: int | None = None,
        gmail_thread_id: str | None = None,
    ) -> AgentResult:
        """Execute the agent loop until completion or max iterations.

        Args:
            profile: Agent profile with system prompt, tools, and model config
            user_message: The initial user message (email content)
            user_id: For logging
            gmail_thread_id: For logging
        """
        messages: list[dict[str, Any]] = []
        if profile.system_prompt:
            messages.append({"role": "system", "content": profile.system_prompt})
        messages.append({"role": "user", "content": user_message})

        tool_specs = self.tool_registry.get_specs(profile.tool_names or None)
        tool_call_records: list[ToolCallRecord] = []

        for iteration in range(1, profile.max_iterations + 1):
            try:
                response = self.llm.agent_completion(
                    messages=messages,
                    tools=tool_specs,
                    model=profile.model,
                    max_tokens=profile.max_tokens,
                    temperature=profile.temperature,
                    user_id=user_id,
                    gmail_thread_id=gmail_thread_id,
                )
            except Exception as e:
                logger.error("Agent loop LLM call failed at iteration %d: %s", iteration, e)
                return AgentResult(
                    status="error",
                    tool_calls=tool_call_records,
                    iterations=iteration,
                    error=str(e),
                )

            message = response.choices[0].message

            # Append assistant message to conversation
            assistant_msg: dict[str, Any] = {"role": "assistant"}
            if message.content:
                assistant_msg["content"] = message.content
            if hasattr(message, "tool_calls") and message.tool_calls:
                assistant_msg["tool_calls"] = [
                    {
                        "id": tc.id,
                        "type": "function",
                        "function": {
                            "name": tc.function.name,
                            "arguments": tc.function.arguments,
                        },
                    }
                    for tc in message.tool_calls
                ]
            messages.append(assistant_msg)

            # If no tool calls, the agent is done
            if not hasattr(message, "tool_calls") or not message.tool_calls:
                return AgentResult(
                    status="completed",
                    final_message=message.content or "",
                    tool_calls=tool_call_records,
                    iterations=iteration,
                )

            # Execute each tool call and feed results back
            for tool_call in message.tool_calls:
                tool_name = tool_call.function.name
                try:
                    arguments = json.loads(tool_call.function.arguments)
                except json.JSONDecodeError:
                    arguments = {}

                logger.info("Agent executing tool %s (iteration %d)", tool_name, iteration)

                result = self.tool_registry.execute(tool_name, arguments)

                record = ToolCallRecord(
                    tool_name=tool_name,
                    arguments=arguments,
                    result=result,
                    iteration=iteration,
                )
                tool_call_records.append(record)

                # Add tool result to conversation
                messages.append(
                    {
                        "role": "tool",
                        "tool_call_id": tool_call.id,
                        "content": json.dumps(result) if not isinstance(result, str) else result,
                    }
                )

        # Exhausted max iterations
        final_content = ""
        for msg in reversed(messages):
            if msg.get("role") == "assistant" and msg.get("content"):
                final_content = msg["content"]
                break

        return AgentResult(
            status="max_iterations",
            final_message=final_content,
            tool_calls=tool_call_records,
            iterations=profile.max_iterations,
        )

File: app/services/agent/loop.rb
# frozen_string_literal: true

module Agent
  # Core agent loop implementing the tool-use pattern.
  # Sends messages to LLM, executes tool calls, and iterates until completion.
  class Loop
    DEFAULT_MAX_ITERATIONS = 10

    def initialize(user:, gmail_client: nil)
      @user = user
      @gmail_client = gmail_client || Gmail::Client.new(user: user)
      @llm = LlmGateway.new(user: user)
    end

    # Execute the agent loop for a given thread.
    # Returns the AgentRun record.
    def execute(gmail_thread_id:, profile_name:, user_message:)
      profile = load_profile(profile_name)
      tool_schemas = Agent::ToolRegistry.instance.tools_for_profile(profile[:tools])

      # Create agent run record
      agent_run = AgentRun.create!(
        user: @user,
        gmail_thread_id: gmail_thread_id,
        profile: profile_name,
        status: "running"
      )

      # Build conversation
      messages = [
        { role: "system", content: profile[:system_prompt] },
        { role: "user", content: user_message }
      ]

      max_iterations = profile[:max_iterations] || DEFAULT_MAX_ITERATIONS
      iteration = 0

      begin
        while iteration < max_iterations
          iteration += 1

          # Call LLM
          response = @llm.chat(
            messages: messages,
            model: profile[:model],
            temperature: profile[:temperature],
            max_tokens: profile[:max_tokens],
            tools: tool_schemas,
            call_type: "agent",
            gmail_thread_id: gmail_thread_id
          )

          # Check if LLM wants to use tools
          if response[:tool_calls].present?
            # Process each tool call
            response[:tool_calls].each do |tool_call|
              tool_name = tool_call[:name]
              tool_args = tool_call[:arguments]

              # Execute tool
              result = Agent::ToolRegistry.instance.execute(
                tool_name,
                arguments: tool_args,
                context: { user: @user, gmail_client: @gmail_client }
              )

              # Log tool call
              agent_run.log_tool_call(
                tool_name: tool_name,
                input: tool_args,
                output: result
              )

              # Add tool call and result to conversation
              messages << { role: "assistant", content: nil, tool_calls: [tool_call] }
              messages << { role: "tool", tool_call_id: tool_call[:id], content: result.to_s }
            end
          else
            # LLM provided final answer
            final_message = response[:content]
            agent_run.complete!(final_message)

            # Log completion event
            email = @user.emails.find_by(gmail_thread_id: gmail_thread_id)
            email&.log_event("agent_completed", "Agent '#{profile_name}' completed: #{final_message&.truncate(200)}")

            return agent_run
          end
        end

        # Max iterations reached
        agent_run.max_iterations!("Reached maximum iterations (#{max_iterations})")
        email = @user.emails.find_by(gmail_thread_id: gmail_thread_id)
        email&.log_event("agent_max_iterations", "Agent '#{profile_name}' hit max iterations (#{max_iterations})")

        agent_run
      rescue StandardError => e
        agent_run.fail!(e.message)
        email = @user.emails.find_by(gmail_thread_id: gmail_thread_id)
        email&.log_event("agent_error", "Agent '#{profile_name}' error: #{e.message}")

        agent_run
      end
    end

    private

    def load_profile(name)
      profiles = AppConfig.get(:agent, :profiles) || {}
      profile_config = profiles[name] || {}

      system_prompt = load_system_prompt(profile_config["system_prompt_file"])

      {
        model: profile_config["model"] || AppConfig.llm.draft_model,
        system_prompt: system_prompt,
        tools: profile_config["tools"] || %w[create_draft escalate],
        temperature: profile_config["temperature"] || 0.3,
        max_tokens: profile_config["max_tokens"] || 4096,
        max_iterations: profile_config["max_iterations"] || DEFAULT_MAX_ITERATIONS
      }
    end

    def load_system_prompt(file_path)
      if file_path.present?
        full_path = Rails.root.join(file_path)
        return File.read(full_path) if File.exist?(full_path)
      end

      # Default agent system prompt
      <<~PROMPT
        You are an AI email assistant agent. Your job is to process incoming emails and take appropriate action.

        Guidelines:
        - Use create_draft for responses that need human review
        - Use escalate for issues beyond your capabilities
        - Be concise and professional in your responses
        - If uncertain, always err on the side of creating a draft rather than auto-sending
      PROMPT
    end
  end
end
